src/util/createValidation.js

```diff
@@ -1,94 +1,41 @@
 import mapValues from 'lodash/mapValues';
 import ValidationError from '../ValidationError';
 import Ref from '../Reference';
-import { asCallback } from './async';
-
-let formatError = ValidationError.formatError;
-
-let thenable = (p) =>
-  p && typeof p.then === 'function' && typeof p.catch === 'function';
-
-function runTest(testFn, ctx, value, sync, callback) {
-  if (sync) {
-    let result,
-      err = null;
-    try {
-      result = testFn.call(ctx, value);
-
-      if (thenable(result)) {
-        throw new Error(
-          `Validation test of type: "${ctx.type}" returned a Promise during a synchronous validate. ` +
-            `This test will finish after the validate call has returned`,
-        );
-      }
-    } catch (e) {
-      err = e;
-    }
-
-    return callback(err, result);
-  }
-
-  return asCallback(
-    new Promise((resolve) => resolve(testFn.call(ctx, value))),
-    callback,
-  );
-}
-
-function resolveParams(oldParams, newParams, resolve) {
-  return mapValues({ ...oldParams, ...newParams }, resolve);
-}
-
-export function createErrorFactory({
-  value,
-  label,
-  resolve,
-  originalValue,
-  ...opts
-}) {
-  return function createError({
-    path = opts.path,
-    message = opts.message,
-    type = opts.name,
-    params,
-  } = {}) {
-    params = {
-      path,
-      value,
-      originalValue,
-      label,
-      ...resolveParams(opts.params, params, resolve),
-    };
-
-    return Object.assign(
-      new ValidationError(formatError(message, params), value, path, type),
-      { params },
-    );
-  };
-}
-
-export default function createValidation(options) {
-  let { name, message, test, params } = options;
 
+export default function createValidation(config) {
```

Comment:
most wins due to simplifying this call by reducing indirection and inlining helpers